src/systemcmds/tests/test_List.cpp

```diff
@@ -0,0 +1,214 @@
+/****************************************************************************
+ *
+ *  Copyright (C) 2019 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file test_List.cpp
+ * Tests flight path smoothing algorithms.
+ */
+
+#include <unit_test.h>
+#include <containers/List.hpp>
+#include <float.h>
+#include <math.h>
+
+class testContainer : public ListNode<testContainer *>
+{
+public:
+	int i{0};
+};
+
+class ListTest : public UnitTest
+{
+public:
+	virtual bool run_tests();
+
+	bool test_add();
+	bool test_remove();
+	bool test_range_based_for();
+
+};
+
+bool ListTest::run_tests()
+{
+	ut_run_test(test_add);
+	ut_run_test(test_remove);
+	ut_run_test(test_range_based_for);
+
+	return (_tests_failed == 0);
+}
+
+bool ListTest::test_add()
+{
+	List<testContainer *> list1;
+
+	// size should be 0 initially
+	ut_compare("size initially 0", list1.size(), 0);
+	ut_assert_true(list1.empty());
+
+	// insert 100
+	for (int i = 0; i < 100; i++) {
+		testContainer *t = new testContainer();
```

Comment:
You do a `new` here. Where is `delete` for it? Does `delete node;` above clean it up correctly?

Reply:
The `clear()` method (https://github.com/PX4/Firmware/blob/4ee8b972599e33ae54d483e5a6949439932afccf/src/include/containers/List.hpp#L138-L148) also deletes all nodes. See line 100 below.

Due to the way this List is used throughout the system we have to be a bit flexible with ownership, which means explicitly calling clear to empty and delete everything and not doing that in the destructor.

For example ModuleParams (https://github.com/PX4/Firmware/blob/f229444722132a2e71d92d425597faaadcb480c4/src/platforms/px4_module_params.h#L98) has a List of all children to update, but doesn't own them.